sdks/java/extensions/arrow/src/main/java/org/apache/beam/sdk/extensions/arrow/ArrowConversion.java

```diff
@@ -0,0 +1,524 @@
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Time type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Timestamp type) {
+                    if (type.getUnit() == TimeUnit.MILLISECOND
+                        || type.getUnit() == TimeUnit.MICROSECOND) {
+                      return FieldType.DATETIME;
+                    } else {
+                      throw new IllegalArgumentException(
+                          "Unsupported timestamp unit: " + type.getUnit().name());
+                    }
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Interval type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Duration type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.LargeBinary type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.LargeUtf8 type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.LargeList type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+                });
+    return fieldType.withNullable(arrowFieldType.isNullable());
+  }
+
+  /**
+   * Returns an {@link Iterable<Row>} backed by the Arrow record batch stored in {@code
+   * vectorSchemaRoot}.
+   *
+   * <p>Note this is a lazy interface. The data in the underlying Arrow buffer is not read until a
+   * field of one of the returned {@link Row}s is accessed.
+   */
+  public static Iterable<Row> rowsFromRecordBatch(
+      Schema schema, VectorSchemaRoot vectorSchemaRoot) {
+    return new RecordBatchIterable(schema, vectorSchemaRoot);
+  }
+
+  public static Iterable<Row> rowsFromRecordBatch(VectorSchemaRoot vectorSchemaRoot) {
+    return rowsFromRecordBatch(
+        ArrowSchemaTranslator.toBeamSchema(vectorSchemaRoot.getSchema()), vectorSchemaRoot);
+  }
+
+  public static VectorSchemaRoot rowFromSerializedRecordBatch(InputStream input)
```

Comment:
nit:
```suggestion
  public static VectorSchemaRoot rowsFromSerializedRecordBatch(InputStream input)
```
Could you modify this to return an `Iterable<Row>` by calling `rowsFromRecordBatch`?

We also need to manage the lifecycle of the `ReadChannel` and `RootAllocator` created in this method, by close()-ing them when done, otherwise we could leak memory. Probably what we should do is make sure the iterator has a reference to these objects, then it can close() them in it's own close method. You could do this either by passing those objects to the Iterable/Iterator, or making the Iterable responsible for creating them.

Then we'll have to call `close()` in `BigQueryStorageArrowReader.close`, and `BigQueryStorageArrowReader.resetBuffer`.

Reply:
I returned the `ReadChannel` and `RootAllocator` to the iterator as it was initially since I think this would be the cleanest approach to closing them, however this implies adding arrow dependencies to java.io.google-cloud-platform. 
The other approach that I was thinking to avoid having those dependencies would be to make a new class in java.extensions.arrow to encapsulate and handle closable resources, would this be a better solution or it is cleaner as it is now?